engine/runtime/src/main/scala/org/enso/compiler/codegen/RuntimeStubsGenerator.scala

```diff
@@ -23,18 +23,27 @@ class RuntimeStubsGenerator(builtins: Builtins) {
       BindingAnalysis,
       "Non-parsed module used in stubs generator"
     )
-    localBindings.constructors.foreach { tp =>
+    val constructors = localBindings.constructors.map { tp =>
       if (tp.builtinType) {
         val builtinType = builtins.getBuiltinType(tp.name)
         if (builtinType == null) {
           throw new CompilerError("Unknown @BuiltinType " + tp.name)
         }
         scope.registerConstructor(builtinType)
         builtinType.setShadowDefinitions(scope)
+        (tp, builtinType)
       } else {
-        val constructor = new AtomConstructor(tp.name, scope)
+        val constructor = new AtomConstructor(tp.name, scope, false)
         scope.registerConstructor(constructor)
+        (tp, constructor)
       }
     }
+
+    // link parents to variants and vice versa
+    constructors.foreach { case (tp,tcons) =>
+      val variants = tp.variants.map(scope.getConstructor(_).get()).toList;
+      tcons.setVariants(variants);
+      variants.foreach(_.setParentType(tcons));
```

Comment:
We can come up with some gymnastics that get us the parent link known to us at initialization time so it can be properly final, but it requires moving a fair bit of stuff around. Usually I avoid circular structures like this, but we need the upward references for the method dispatch (eventually) and the downward links for hinting in the UI.

Another thing I carefully considered (and implemented first) was to separate those two views into two separate types, but it led to a lot of useless type duplication, lots of noise ordering things. etc.

Reply:
Seems like a plan.